_posts/2023-05-10-group-equivariant-cnn-3/group-equivariant-cnn-3.Rmd

```diff
@@ -0,0 +1,492 @@
+---
+title: "Group-equivariant neural networks with escnn"
+description: >
+  Escnn, built on PyTorch, is a library that, in the spirit of Geometric Deep Learning, provides a high-level interface to designing and training group-equivariant neural networks. This post introduces important mathematical concepts, the library's key actors, and essential library use.
+author:
+  - name: Sigrid Keydana
+    affiliation: Posit
+    affiliation_url: https://www.posit.co/
+slug: keydanagcnn3
+date: 2023-05-10
+categories:
+  - Torch
+  - R
+  - Concepts
+  - Image Recognition & Image Processing
+bibliography: references.bib
+output:
+  distill::distill_article:
+    self_contained: false
+    toc: true
+preview: images/preview.jpg
+---
+
+```{r setup, include=FALSE}
+knitr::opts_chunk$set(echo = TRUE, eval = FALSE, fig.width = 6, fig.height = 6)
+```
+
+Today, we resume our exploration of group equivariance. This is the third post in the series. The [first](https://blogs.rstudio.com/ai/posts/2023-03-09-group-equivariant-cnn-1/) was a high-level introduction: What this is all about; how equivariance is operationalized; and why it is of relevance to many deep-learning applications. The [second](https://blogs.rstudio.com/ai/posts/2023-03-27-group-equivariant-cnn-2/) sought to concretize the key ideas by developing a group-equivariant CNN from scratch. That being instructive, but too tedious for practical use, today we look at a carefully designed, highly-performant [library](https://github.com/QUVA-Lab/escnn) that hides the technicalities and enables a convenient workflow.
+
+First though, let me again set the context. In physics, an all-important concept is that of symmetry[^1], a symmetry being present whenever some quantity is being conserved. But we don't even need to look to science. Examples arise in daily life, and -- otherwise why write about it - in the tasks we apply deep learning to. In daily life: Think about speech -- me stating "it is cold", for example. Formally, or denotation-wise, the sentence will have the same meaning now as in five hours. (Connotations, on the other hand, can and will probably be different!). This is a form of translation symmetry, translation in time. In deep learning: Take image classification. For the usual convolutional neural network, a cat in the center of the image is just that, a cat; a cat on the bottom is, too. But one sleeping, comfortably curled like a half-moon "open to the right", will not be "the same" as one in a mirrored position. Of course, we can train the network to treat both as equivalent by providing training images of cats in both positions, but that is not a scaleable approach. Instead, we'd like to make the network aware of these symmetries, so they are automatically preserved throughout the network architecture.
```

Comment:
> In daily life: Think about speech -- me stating "it is cold", for example.

This is a little off structurally. I think, since you've already mentioned that there are examples from daily life, maybe you could cut that. Otherwise, I'd integrate it into the sentence rather than separating it into a colon-delimited introduction.

Reply:
> In deep learning: Take image classification.

OK, now I kind of see what you're doing here, but it was a bit hard to follow. I think rather than have this sentence introducing both everyday-life and deep-learning examples, I'd give each example a fuller introduction than, e.g. "In deep learning." That, or separate them into paragraphs.

Reply:
Yes, I see what you mean. I was thinking about having separate paragraphs, but then deciding against it, for the usual bad reasons ... (_doesn't seem important enough_)